tests/utils/test_shell_util.py

```diff
@@ -426,49 +419,54 @@ def invoke(command):
-                    raise Exception("The child processes did not complete within the allowed timeout")
+                if not self._wait_for(all_commands_running):
+                    self.fail("shellutil.get_running_commands() did not report the expected number of commands after the allowed timeout. Got: {0}".format(
+                        self._format_pids(running_commands[0])))
 
                 started_commands.sort()
-                started_commands_expected.sort()
+                running_commands[0].sort()
+
                 self.assertEqual(
-                    started_commands_expected,
                     started_commands,
-                    "The command_started callback was not invoked with the correct processes.\nExpected: {0}\nGot: {1}".format(
-                        [self._get_command_line(pid) for pid in started_commands_expected],
-                        [self._get_command_line(pid) for pid in started_commands]))
+                    running_commands[0],
+                    "shellutil.get_running_commands() did not return the expected commands.\nExpected: {0}\nGot: {1}".format(
+                        self._format_pids(started_commands), self._format_pids(running_commands[0])))
 
             finally:
                 # terminate the child processes, since they are blocked
                 for pid in child_processes:
-                    try:
-                        os.kill(pid, signal.SIGKILL)
-                    except Exception:
-                        pass  # no need to handle this error
-
-            # wait for the command_completed callbacks, and verify them
-            if not self._wait_for(lambda: len(completed_commands) == len(started_commands)):
-                raise Exception("The child processes did not complete within the allowed timeout")
-
-            completed_commands.sort()
-            self.assertEqual(
-                started_commands,
-                completed_commands,
-                "The command_completed callback does not match the invocations of command_started.\nStarted: {0}\nCompleted: {1}".format(
-                    [self._get_command_line(pid) for pid in started_commands],
-                    [self._get_command_line(pid) for pid in completed_commands]))
-        finally:
-            shellutil.set_on_command_started_callback(None)
-            shellutil.set_on_command_completed_callback(None)
+                    os.kill(pid, signal.SIGKILL)
 
+            # once the processes complete, their PIDs should go away
+            running_commands = [[]]
+
+            def all_commands_completed():
+                running_commands[0] = shellutil.get_running_commands()
+                return len(running_commands[0]) == 0
+
+            if not self._wait_for(all_commands_completed):
+                self.fail("shellutil.get_running_commands() should return empty after the commands complete. Got: {0}".format(
+                    self._format_pids(running_commands[0])))
```

Comment:
Would it be useful for debugging to output the process names along with PIDs?

Reply:
_format_pids does that

Reply:
You're right, based on the name I assumed it just formats the PIDs, not the names too.